{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Methylation Array CNV Analysis\n",
    "\n",
    "**Author** - Clarence Mah\n",
    "<br>\n",
    "**Email** - ckmah@ucsd.edu\n",
    "\n",
    "This notebook performs copy number variation (CNV) analysis on the methylation profiles of several glioblastoma tumor samples taken from a comprehensive [CNS tumor study (Capper et al. 2018)](https://www.nature.com/articles/nature26000).\n",
    "\n",
    "### Objective\n",
    "The goal of this notebook is to provide a workflow for pre-processing raw array data to copy number variation (CNV) analysis. Analyses performed in this workflow use the [minfi](http://bioconductor.org/packages/release/bioc/html/minfi.html) and [conumee](http://bioconductor.org/packages/release/bioc/html/conumee.html) packages.\n",
    "\n",
    "### Dataset\n",
    "The data used in this notebook can be found on [GEO](https://www.ncbi.nlm.nih.gov/geo/), accession number: [GSE90496](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE90496). The sample data is 450k methylation array data of 2 glioblastoma tumor samples, while the control samples for copy number analysis normalization include 450k methylation array data of 119 normal brain tissue samples.\n",
    "\n",
    "### Analysis Overview\n",
    "1. **[Load & Preprocess Data](#Step-1:-Load-&-Preprocess-Data)**\n",
    "    1. Load raw sample files and output a `MethylSet` file.\n",
    "    2. Load raw control sample files and output a `MethylSet` file.\n",
    "2. **[Copy Number Variation Analysis](#Step-2:-Copy-Number-Variation-Analysis)**\n",
    "    1. Generate copy number calls and plots. \n",
    "\n",
    "# Workflow Diagram\n",
    "![Flow diagram](cnv_flow_diagram.svg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Login to GenePattern\n",
    "Login to a GenePattern server to perform the following analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T15:30:24.025793Z",
     "start_time": "2018-05-15T15:30:24.015323Z"
    },
    "genepattern": {
     "server": "https://gp-beta-ami.genepattern.org/gp",
     "type": "auth"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0e952b69069b43fd9fd4bc304e9255de",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>GPAuthWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "GPAuthWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Requires GenePattern Notebook: pip install genepattern-notebook\n",
    "import gp\n",
    "import genepattern\n",
    "\n",
    "# Username and password removed for security reasons.\n",
    "genepattern.GPAuthWidget(genepattern.register_session(\"https://gp-beta-ami.genepattern.org/gp\", \"\", \"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1: Load & Preprocess Data\n",
    "\n",
    "This step accepts an archive of raw `IDAT` files and consolidates the intensity values into a single file.\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n",
    "<p>**Supported file formats**: `zip`, `gz`, and `.tar.gz`.</p>\n",
    "\n",
    "[**Step 1.1**](#Step-1.1:-Sample-data) - Load and preprocess your sample data.<br>\n",
    "[**Step 1.2**](#Step-1.2:-Control-data) - If your control samples are part of your sample data, skip this step and specify the control samples names in [**Step 2**](#Step-2:-Copy-Number-Variation-Analysis). Otherwise, load and preprocess your control data.\n",
    "\n",
    "</div>\n",
    "### Archive Folder Structure\n",
    "The `IDAT` files can be organized one of 2 ways: either as a flat archive - **OR** - archived in the following folder structure described in the [Illumina Demo Dataset](ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/productfiles/methylationEPIC/infinium-methylationepic-demo-dataset.zip).\n",
    "\n",
    " **Flat archive**:\n",
    "- `dataset.zip`\n",
    "    - 200144450018_R04C01_1_Green.xml\n",
    "    - 200144450018_R04C01_1_Red.xml\n",
    "    - 200144450018_R04C01_2_Green.xml\n",
    "    - 200144450018_R04C01_2_Red.xml\n",
    "    - ...\n",
    "\n",
    " **Illumina output structure**:\n",
    "- `dataset.zip`\n",
    "    - Demo_SampleSheet.csv\n",
    "    - 200144450018\n",
    "        - 200144450018_R04C01_1_Green.xml\n",
    "        - 200144450018_R04C01_1_Red.xml\n",
    "        - 200144450018_R04C01_2_Green.xml\n",
    "        - 200144450018_R04C01_2_Red.xml\n",
    "        - ...\n",
    "    - 200144450019\n",
    "        - ...\n",
    "    - ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.1: Sample data\n",
    "\n",
    "Load and preprocess the sample data.\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n",
    "<ol>\n",
    "    <li>Load your sample **dataset** as one of the supported file formats. A detailed example of the folder structure is [provided above](#Input-folder-structure).</li>\n",
    "    <li>Set **normalization** to `None` (Quantile and functional normalizations also performs scaling, which will not work with **Step 2**).</li>\n",
    "    <li> Set **output type** to `MethylSet`.</li>\n",
    "</ol>\n",
    "\n",
    "<p>Example sample data (2 glioblastoma samples): Drag ftp://gpftp.broadinstitute.org/methylation/gbm_sample.tar.gz to the **dataset** field.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T15:30:24.680148Z",
     "start_time": "2018-05-15T15:30:24.669052Z"
    },
    "genepattern": {
     "type": "task"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1076018b8e774636b4e91ba1689fdaec",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>GPTaskWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "GPTaskWidget(lsid='urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00373')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "minfipreprocessing_task = gp.GPTask(genepattern.get_session(0), 'urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00373')\n",
    "minfipreprocessing_job_spec = minfipreprocessing_task.make_job_spec()\n",
    "minfipreprocessing_job_spec.set_parameter(\"dataset\", \"ftp://gpftp.broadinstitute.org/methylation/gbm_sample.tar.gz\")\n",
    "minfipreprocessing_job_spec.set_parameter(\"normalization\", \"None\")\n",
    "minfipreprocessing_job_spec.set_parameter(\"output.type\", \"MethylSet\")\n",
    "genepattern.GPTaskWidget(minfipreprocessing_task)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T15:30:24.695306Z",
     "start_time": "2018-05-15T15:30:24.682423Z"
    },
    "genepattern": {
     "type": "job"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "be4f102cafce43d5a3b4f0a865b9ffb5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>GPJobWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "GPJobWidget(job_number=24005)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "job24005 = gp.GPJob(genepattern.get_session(0), 24005)\n",
    "genepattern.GPJobWidget(job24005)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.2: Control data\n",
    "Load and preprocess the controls sample data.\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n",
    "<ol>\n",
    "    <li>Load your control **dataset** as one of the supported file formats. A detailed example of the folder structure is [provided above](#Input-folder-structure).</li>\n",
    "    <li>Set **normalization** to `None` (Quantile and functional normalizations also performs scaling, which will not work with **Step 2**).</li>\n",
    "    <li> Set **output type** to `MethylSet`.</li>\n",
    "</ol>\n",
    "\n",
    "<p>Example control data (119 normal brain tissue samples): Drag ftp://gpftp.broadinstitute.org/methylation/controls.tar.gz to the **dataset** field.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T15:30:24.711718Z",
     "start_time": "2018-05-15T15:30:24.698323Z"
    },
    "genepattern": {
     "show_code": false,
     "type": "task"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7b3c2a18dc4c45518caf2618c981579b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>GPTaskWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "GPTaskWidget(lsid='urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00373')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "minfipreprocessing_task = gp.GPTask(genepattern.get_session(0), 'urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00373')\n",
    "minfipreprocessing_job_spec = minfipreprocessing_task.make_job_spec()\n",
    "minfipreprocessing_job_spec.set_parameter(\"dataset\", \"ftp://gpftp.broadinstitute.org/methylation/controls.tar.gz\")\n",
    "minfipreprocessing_job_spec.set_parameter(\"normalization\", \"None\")\n",
    "minfipreprocessing_job_spec.set_parameter(\"output.type\", \"MethylSet\")\n",
    "genepattern.GPTaskWidget(minfipreprocessing_task)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T15:30:24.728416Z",
     "start_time": "2018-05-15T15:30:24.713891Z"
    },
    "genepattern": {
     "show_code": true,
     "type": "job"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b73cfc554b1d436f9ad4fadfd27d734d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>GPJobWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "GPJobWidget(job_number=24004)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "job24004 = gp.GPJob(genepattern.get_session(0), 24004)\n",
    "genepattern.GPJobWidget(job24004)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 2: Copy Number Variation Analysis\n",
    "\n",
    "This step accepts unscaled intensity values. The CNV analysis algorithm is roughly as follows (adapted from [conumee vignette](https://bioconductor.org/packages/release/bioc/vignettes/conumee/inst/doc/conumee.html#perform-cnv-analysis)):\n",
    ">1. **Per sample normalization**: Each sample is normalized to the control samples by multiple linear regression. The regression analysis yields the linear combination of control samples that most closely fits the intensities of the query sample. Subsequently, the log2 ratio of probe intensities of the query sample versus the combination of control samples are calculated and used for further analysis.\n",
    "2. **Probes are grouped by proximity**: Probes are combined within predefined genomic bins. Intensity values are shifted to minimize the median absolute deviation from all  bins to zero to determine the copy-number neutral state.\n",
    "3. **Highlight genes**: The regions for each `gene to highlight` are queried from Ensembl database (hg19) with `BiomaRt`.\n",
    "4. **Genome segmentation**: Finally, the genome is segmented into regions of the same copy number state.\n",
    "<div class=\"alert alert-info\">\n",
    "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n",
    "<ol>\n",
    "    <li>Click on the **sample data** parameter text input. Select your sample `methyl_set.rds` file from the dropdown.\n",
    "    </li>\n",
    "    <li>If your control samples are in your sample data, specify their names in the **control sample names** parameter. Otherwise, load the `methyl_set.rds` from **Step 1.2** the same way **sample data** is loaded.</li>\n",
    "    <li>Set the appropriate **array type**. Choose `450k` for the example data.</li>\n",
    "    <li>Set **genes to highlight** to a file with your genes of interest.</li>\n",
    "    <li>Set **exclude regions** to a bed file specifying genomic regions to exclude from copy number analysis.</li>\n",
    "    <li>Choose **sex chromosomes**. Choose `Yes` for the example dataset.</li>\n",
    "</ol>\n",
    "Example **genes to highlight** gene list: ftp://gpftp.broadinstitute.org/methylation/detail_genes.txt<br>\n",
    "<ul style=\"margin: 0\"><li>List of genes that are known frequently mutated genes across all brain tumors + glioblastoma specific genes.</li>\n",
    "</ul><br>\n",
    "Example **exclude regions** bed file: ftp://gpftp.broadinstitute.org/methylation/exclude_regions.bed\n",
    "<ul style=\"margin: 0\"><li>List of highly polymorphic regions specified by the `conumee` package.</li>\n",
    "</ul>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T15:30:24.751206Z",
     "start_time": "2018-05-15T15:30:24.731018Z"
    },
    "genepattern": {
     "type": "task"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "889feab8912f4924a6e836e266b36bb4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>GPTaskWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "GPTaskWidget(lsid='urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00375')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "conumeecnvanalysis_task = gp.GPTask(genepattern.get_session(0), 'urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00375')\n",
    "conumeecnvanalysis_job_spec = conumeecnvanalysis_task.make_job_spec()\n",
    "conumeecnvanalysis_job_spec.set_parameter(\"sample.data\", \"\")\n",
    "conumeecnvanalysis_job_spec.set_parameter(\"control.sample.names\", \"\")\n",
    "conumeecnvanalysis_job_spec.set_parameter(\"control.samples.data\", \"https://gp-beta-ami.genepattern.org/gp/jobResults/24004/methyl_set.rds\")\n",
    "conumeecnvanalysis_job_spec.set_parameter(\"array.type\", \"450k\")\n",
    "conumeecnvanalysis_job_spec.set_parameter(\"genes.to.highlight\", \"ftp://gpftp.broadinstitute.org/methylation/detail_genes.txt\")\n",
    "conumeecnvanalysis_job_spec.set_parameter(\"exclude.regions\", \" ftp://gpftp.broadinstitute.org/methylation/exclude_regions.bed\")\n",
    "conumeecnvanalysis_job_spec.set_parameter(\"sex.chromosomes\", \"Yes\")\n",
    "genepattern.GPTaskWidget(conumeecnvanalysis_task)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T15:30:24.768288Z",
     "start_time": "2018-05-15T15:30:24.754021Z"
    },
    "genepattern": {
     "type": "job"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a1549eacc74d451aa4e56c8e9d6b1740",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>GPJobWidget</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "GPJobWidget(job_number=24209)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "job24209 = gp.GPJob(genepattern.get_session(0), 24209)\n",
    "genepattern.GPJobWidget(job24209)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
